File: lib/task1.ml
let triangle_number n = n * (n + 1) / 2

let count_divisors n =
  let rec count_divisors_rec i count =
    if i * i > n then count
    else if n mod i = 0 then
      if i * i = n then count_divisors_rec (i + 1) (count + 1)
      else count_divisors_rec (i + 1) (count + 2)
    else count_divisors_rec (i + 1) count
  in
  count_divisors_rec 1 0

(* 1. Монолитная реализация с хвостовой рекурсией *)
let solve1_tail_rec min_divisors =
  let rec find_triangle n =
    let tri = triangle_number n in
    let divisors = count_divisors tri in
    if divisors > min_divisors then tri else find_triangle (n + 1)
  in
  find_triangle 1

(* 1. Монолитная реализация с обычной рекурсией *)
let solve1_rec min_divisors =
  let rec find_triangle n =
    let tri = triangle_number n in
    let divisors = count_divisors tri in
    if divisors > min_divisors then tri
    else
      let next = find_triangle (n + 1) in
      next
  in
  find_triangle 1

let solve1_modular min_divisors =
  let generate_triangles n =
    let rec gen acc i =
      if i > n then List.rev acc else gen (triangle_number i :: acc) (i + 1)
    in
    gen [] 1
  in

  let filter_sufficient_divisors tri_list =
    List.filter (fun tri -> count_divisors tri > min_divisors) tri_list
  in

  let find_first = function
    | [] -> failwith "No triangle number found"
    | x :: _ -> x
  in

  let limit = 20000 in
  generate_triangles limit |> filter_sufficient_divisors |> find_first

(* 3. Генерация последовательности при помощи map *)
let solve1_map min_divisors =
  let limit = 20000 in
  let indices = List.init limit (fun i -> i + 1) in
  let triangles = List.map triangle_number indices in
  let with_divisors =
    List.map (fun tri -> (tri, count_divisors tri)) triangles
  in
  let filtered =
    List.filter (fun (_, divs) -> divs > min_divisors) with_divisors
  in
  match filtered with
  | [] -> failwith "No triangle number found"
  | (tri, _) :: _ -> tri

(* 4. Работа со спец. синтаксисом для циклов (for) *)
let solve1_for_loop min_divisors =
  let result = ref None in
  let found = ref false in
  for n = 1 to 20000 do
    if not !found then
      let tri = triangle_number n in
      let divisors = count_divisors tri in
      if divisors > min_divisors then (
        result := Some tri;
        found := true)
  done;
  match !result with Some x -> x | None -> failwith "No triangle number found"

(* 5. Работа с бесконечными последовательностями (Seq) *)
let solve1_seq min_divisors =
  let rec ints_from n = fun () -> Seq.Cons (n, ints_from (n + 1)) in
  let triangles = Seq.map triangle_number (ints_from 1) in
  let with_divisors =
    Seq.map (fun tri -> (tri, count_divisors tri)) triangles
  in
  let filtered =
    Seq.filter (fun (_, divs) -> divs > min_divisors) with_divisors
  in
  match filtered () with
  | Seq.Cons ((tri, _), _) -> tri
  | Seq.Nil -> failwith "No triangle number found"

File: python/task1.py
"""
Задача 1: Найти первое треугольное число с более чем 500 делителями
Реализация на Python для сравнения
"""

import time
import math


def triangle_number(n):
    """Вычисление n-го треугольного числа"""
    return n * (n + 1) // 2


def count_divisors(n):
    """Подсчёт количества делителей числа"""
    count = 0
    sqrt_n = int(math.sqrt(n))
    for i in range(1, sqrt_n + 1):
        if n % i == 0:
            if i * i == n:
                count += 1
            else:
                count += 2
    return count


def solve1_imperative(min_divisors):
    """Императивная реализация с циклом while"""
    n = 1
    while True:
        tri = triangle_number(n)
        divisors = count_divisors(tri)
        if divisors > min_divisors:
            return tri
        n += 1


def solve1_for_loop(min_divisors):
    """Реализация с циклом for"""
    for n in range(1, 20000):
        tri = triangle_number(n)
        divisors = count_divisors(tri)
        if divisors > min_divisors:
            return tri
    return None


def solve1_functional(min_divisors):
    """Функциональная реализация с генераторами"""
    def triangle_numbers():
        n = 1
        while True:
            yield triangle_number(n)
            n += 1
    
    def filter_divisors(triangles):
        for tri in triangles:
            if count_divisors(tri) > min_divisors:
                yield tri
    
    return next(filter_divisors(triangle_numbers()))


def solve1_list_comprehension(min_divisors):
    """Реализация с list comprehension"""
    limit = 20000
    triangles = [triangle_number(n) for n in range(1, limit + 1)]
    with_divisors = [(tri, count_divisors(tri)) for tri in triangles]
    filtered = [tri for tri, divs in with_divisors if divs > min_divisors]
    return filtered[0] if filtered else None


if __name__ == "__main__":
    min_divisors = 500
    
    print("=== Задача 1: Треугольные числа с более чем 500 делителями ===")
    print()
    
    print("1. Императивная реализация (while):")
    start_time = time.time()
    result1 = solve1_imperative(min_divisors)
    end_time = time.time()
    print(f"   Результат: {result1} (время: {end_time - start_time:.4f} сек)")
    print()
    
    print("2. Реализация с for-циклом:")
    start_time = time.time()
    result2 = solve1_for_loop(min_divisors)
    end_time = time.time()
    print(f"   Результат: {result2} (время: {end_time - start_time:.4f} сек)")
    print()
    
    print("3. Функциональная реализация (генераторы):")
    start_time = time.time()
    result3 = solve1_functional(min_divisors)
    end_time = time.time()
    print(f"   Результат: {result3} (время: {end_time - start_time:.4f} сек)")
    print()
    
    print("4. Реализация с list comprehension:")
    start_time = time.time()
    result4 = solve1_list_comprehension(min_divisors)
    end_time = time.time()
    print(f"   Результат: {result4} (время: {end_time - start_time:.4f} сек)")
    print()

